README.md
# Events API

This repository contains the Events API. The Events API allows authenticated users to create, view, and mark their RSVP status to events for their community.

This API is fully RESTful.


## Authentication

All requests to the API must be authenticated, except for sign in/up requests. This API follows the OAuth 2.0 specification. Access Tokens must be passed through the `Authorization` header on any authenticated endpoints in the following format:

```
Authorization: Bearer {Token}
```


## API Anatomy

The Events API can be broken into a few core resources and their methods:

- User
	- `GET` my user
	- `GET` another user
- Event
	- `GET` all events
	- `GET` an event (by ID)
	- `POST` an event
	- `PUT` an event
	- `DELETE` an event
	- `GET` an event's attendees
- RSVP
	- `POST` an RSVP to an event
	- `DELETE` an RSVP to an event
	- `PUT` an RSVP to an event

## Responses

All response headers MUST follow the [HTTP/1.1 status codes definitions](https://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html). 

All response bodies MUST be JSON-encoded, and fully parsable by any JSON serializer.

At a fundamental level, there are two Events API response types:

* [Success response](#success-response)
* [Error response](#error-response)

All resource endpoints MUST use one of these response formats.

## Success Response

To qualify a success, the response MUST have a `2xx`/`3xx`-level HTTP status code, MUST NOT have an `error` field, and MAY include a `data` field. e.g.:

```
GET -> /v1/users/me
```
--
```
200 OK
```
```
{
    "data": {
        "id": 5,
        "type": "user",
        "name": "jeremy",
        "bio": "loves to host big house parties"
    }
}
```

*Note: Some success responses may include additional fields, such as the `meta` field.*

## Error Response

To qualify an error, the response MUST have a `4xx`/`5xx`-level HTTP status code, and MUST include the `error` field which consists of an [error object](Errors.md#error-object). E.g.:

```
GET -> /v1/users/me/
```
--
```
401 Unauthorized
```
```
{
    "error": {
        "message": "Your request could not be completed because it was not properly authenticated.",
        "code": 300
    }
}
```

## Errors

The Events API returns a number of different errors. The HTTP response status code will be in accordance with the [W3 Status Code Definitions](https://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html); however, the response data will include internal status codes which are detailed below:

Code | Message | HTTP code | Description | Reference
---- | ------- | --------- | ----------- | ---------  
300 | Your request could not be completed because it was not properly authenticated. | 401 | Something was wrong with the authentication method for this request. | [Request Authentication](Authentication/RequestAuthentication.md)
301 | A valid access token is required to access this resource. | 401 | | [Request Authentication](Authentication/RequestAuthentication.md)
302 | You must be authenticated as a user to take this action. | 401 | Identity required to make this request. | [User Authentication](Authentication/OAuthAuthentication.md), [Request Authentication](Authentication/RequestAuthentication.md)
310 | The provided access token is invalid, or cannot be used with the provided parameters. | 401 | Invalid or expired [Access Token](Authentication/OAuthAuthentication/AccessToken.md) provided. | [User Authentication](Authentication/OAuthAuthentication.md), [Access Token](Authentication/OAuthAuthentication/AccessToken.md)
311 | The provided refresh token is invalid, or cannot be used with the provided parameters. | 400 | Invalid Refresh Token provided. | [User Authentication](Authentication/OAuthAuthentication.md), [Access Token](Authentication/OAuthAuthentication/AccessToken.md)
312 | The provided access token or refresh token does not match the given bearer token. | 401 | Mismatched API Key provided when refreshing token. | [User Authentication](Authentication/OAuthAuthentication.md)
320 | The provided refresh token cannot be used to renew an access token. Client login should be attempted. | 403 | | [User Authentication](Authentication/OAuthAuthentication.md)
700 | The requested operation is not permitted on this resource. | 412 | A precondition was not met for this request. | 
701 | Invalid request params | 412 | | 
23505 | Key (username or email)=(value) already exists | 409 | A duplicate username or email was provided in this request |


changelog.md
# Changelog
All commits are documented in this file.

## [07/20/2020]
### Added 
 - frontend validation for sign in 

## [07/03/2020]
### Changed
 - Moved `create()` and `retrieve()` out of User class
 - small variable changes on frontend
 - apiError.output function

## [06/30/2020]
### Changed
 - moved `users.js` functions into the `User` class (unfinished, throws an error)

## [06/28/2020]
### Added
 - User class
 - `users.create()` returns a `new User`
 
### Changed
 - apiError throwing in `users.create()` (adding user class messed it up)
 - frontend to recieve new User object

## [06/27/2020]
### Added
 - error lookup in `apiError.js` (given only the error code)
 - frontend error handling for duplicates 
### Removed
 - utf-8 encoding and decoding in `events.js` and `users.js`

## [06/24/2020]
### Added
 - Error handling for unique constraint violation

## [06/23/2020]
### Added
 - Unique constraint to username and email columns in DB

## [06/22/2020]
### Added
 - `validate()` and `format()` for `events` model
 - UTF-8 encoding for trimmed data (before adding to DB)
 - decoding data in `users.format()` before sending it to frontend
 - more specific data validation to password and email in `users` model
 - checking if error is apiError before running through the error handler

### Changed
- Changed frontend date JS to create the correct timestamp

## [06/20/2020]
### Added
 - function to check if db is up 

### Changed 
 - endpoints to just `user` and `event`

### Removed
 - old error handling for `event` endpoint 

## [06/19/2020]
### Added
 - apiError class 
 - better error handling

## [06/17/2020]
### Added
 - upcoming function to events model to return upcoming events
 - timestamp column
 - frontend code to create timestamp (with timezone) from input data (events form)

### Changed
 - Error handling in model functions to promises (.catch())

### Removed
 - date and time columns

## [06/16/2020]
### Added
 - function in users model that creates a data object for frontend
 - personalize endpoints for redirect after sign up
 - sign in form 

### Changed 
 - moved events form to user page

## [06/15/2020]
### Added 
 - frontend validation 
 - error handling for backend validation
 - cookie for userID 
 - redirect to page that gets user data (with userId cookie)

## [06/14/2020]
### Added 
 - basic data validation

## [06/12/2020]
### Added
 - styling for creating user
 - create event form
 - backend to receive event data from frontend and add it to DB
 - select event from DB and send it to frontend
 - 3 separate models for users, events, and rsvps
 - retrieve(), update(), and delete() for 3 models

### Changed
 - Date and Time datatypes in event table 
 - migrated query functions over to 3 models

### Removed
 - linked_username and linked_event_name columns from RSVP
 - queries model

## [06/11/2020]
### Added 
 - RSVP table in DB
 - functions to add users, events, and rsvps
 - frontend from to collect userdata
 - post request sending userdata
 - backend receiving userdata
 - function to select user table from DB
 - send user table to frontend

## [06/10/2020]
### Added 
 - postgres DB and express environment
 - basic tables for api
 - changelog

[06/30/2020]: https://github.com/Gbillington1/Events-API/compare/7d89bac..HEAD
[06/28/2020]: https://github.com/Gbillington1/Events-API/compare/1fe593d..a30b95c
[06/27/2020]: https://github.com/Gbillington1/Events-API/compare/d6fbea3..83cd0e4
[06/24/2020]: https://github.com/Gbillington1/Events-API/compare/c064bfc..2245766
[06/23/2020]: https://github.com/Gbillington1/Events-API/compare/e12d10f..95740b2
[06/22/2020]: https://github.com/Gbillington1/Events-API/compare/f366b6c..fadd225
[06/20/2020]: https://github.com/Gbillington1/Events-API/compare/97a7068..1d52998
[06/19/2020]: https://github.com/Gbillington1/Events-API/compare/157fe63..c19ad2c
[06/17/2020]: https://github.com/Gbillington1/Events-API/compare/3087dde..53d97b7
[06/16/2020]: https://github.com/Gbillington1/Events-API/compare/c0d2478..749cb2c
[06/15/2020]: https://github.com/Gbillington1/Events-API/compare/68239d0..bb085bb
[06/14/2020]: https://github.com/Gbillington1/Events-API/compare/34e1f87..34e1f87
[06/12/2020]: https://github.com/Gbillington1/Events-API/compare/0c7c2df..edfc0e1
[06/11/2020]: https://github.com/Gbillington1/Events-API/compare/277c5bf..76c0318
[06/10/2020]: https://github.com/Gbillington1/Events-API/compare/d3b101d..c0e50d9

src/app.js
console.log('yeet')
const express = require('express');
const app = express();
const port = parseInt(process.env.EXPOSED_PORT);
const fs = require('fs');
const { Client } = require('pg');
const cookieParser = require('cookie-parser');
const users = require('./models/users');
const events = require('./models/events');
const rsvps = require('./models/rsvps');
const apiError = require("./errors/apiError");
let isDbUp = false;
var currentUser;

// use cookie-parsing middleware
app.use(cookieParser());

app.use(express.urlencoded({ extended: true }));

// basic validation
app.use(function (req, res, next) {
    var data;
    switch (req.method) {
        case "GET":
            data = req.query
            break;

        default:
            data = req.body;
            break;
    }
    var values = Object.values(data);
    var keys = Object.keys(data);
    for (var i = 0; i < values.length; i++) {
        if (typeof values[i] == typeof undefined || values[i] == "") {
            var err = new apiError(701, "Invalid request params at " + keys[i]);
            next(err);
            return;
        }
    }
    next();
})
console.log('bruh')
//connect to DB 
const client = new Client({
    connectionString: "postgres://root:pass@postgres:5432/events"
});

client
    .connect()
    .then(() => {
        console.log('connected')
        isDbUp = true;
    })
    .catch(err => console.error('error', err.stack));

function checkDb() {
    if (isDbUp) {
        return client;
    }
}

//serve html homepage
app.get('/', function (req, res) {
    fs.readFile("./index.html", function (err, data) {
        if (err) {
            res.writeHead(404);
            res.write("Not found");
            res.end();
        } else {
            res.writeHead(200, { 'Content-Type': 'text/html' });
            res.write(data);
            res.end();
        }
    });
});

app.get('/user/:username', function (req, res) {
    fs.readFile('./userIndex.html', function (err, data) {
        if (err) {
            res.writeHead(404);
            res.write("Not found");
            res.end();
        } else {
            res.writeHead(200, { 'Content-Type': 'text/html' });
            res.write(data);
            res.end();
        }
    })
})

app.get('/user', function (req, res) {

    users.retrieve(checkDb(), req.query.userId).then(function (response) {

        var currentUser = response;

        res.send(currentUser.output())

    }).catch(err => { console.error(err) })
})

// receive post request to /user endpoint
app.post('/user', function (req, res, next) {
    // add data from form to userData obj
    var userData = users.validate(req.body);

    // add userData to DB => return new User
    users.create(checkDb(), userData).then(function (data) {

        currentUser = data;
        res.cookie('userId', data.id);
        res.cookie('username', data.username);
        res.send(currentUser.output());


    }).catch(err => {

        next(err)

    });
})

app.get('/event', function (req, res) {
    events.upcoming(checkDb()).then(function (rows) {
        if (rows.length > 0) {
            var data = {};
            for (var i = 0; i < rows.length; i++) {
                data[i] = events.format(rows[i])
            }
            res.send(data)
        } else {
            res.end();
        }

    }).catch(err => { console.error(err) })
})

// reveive post request to /addEvent endpoint
app.post('/event', function (req, res) {
    // form eventData with data form frontend
    console.log(req.body)
    var eventData = events.validate(req.body);
    console.log(eventData)
    // add event to DB
    events.create(checkDb(), eventData);

    // get event from DB and send it to frontend
    events.all(checkDb()).then(function (rows, error) {
        if (typeof error !== "undefined") {
            res.send({ 'error': true });

            return;
        }

        console.log(rows[rows.length - 1])
        var data = events.format(rows[rows.length - 1]);
        console.log(data)
        res.send(data);
    }).catch(err => { console.error(err) })
});

// error handling
app.use(function (err, req, res, next) {
    var httpCode;
    if (err instanceof apiError) {

        httpCode = err.httpCode
        err = err.output();
        res.status(httpCode);

    } else {

        httpCode = err.code;

    }

    console.error(err)
    res.send(err);
    
})

// listen for connection on localhost
app.listen(port, () => console.log(`App listening at http://localhost:${port}`))


src/models/rsvps.js
// havent tested any of these

// create an rsvp
function create(client, data) {
    client.query('INSERT INTO rsvps (linked_event_id, linked_user_id, rsvp_status) VALUES ($1, $2, $3)', [data.eventID, data.userID, data.rsvpStatus]);
}

// return rsvp with specific event and user id
function retrieve(client, eventID, userID) {
    return new Promise(function (resolve, reject) {
        client.query('SELECT * FROM rsvps WHERE linked_event_id = $1 AND linked_user_id = $2', [eventID, userID], function (err, response) {
            resolve(response.rows)
        })
    })
}

// update rsvp status with specific event and user id
function update(client, data) {
    client.query('UPDATE rsvps SET rsvp_status = $1 WHERE linked_event_id = $2 AND linked_user_id = $3', [data.rsvpStatus, data.eventID, data.userID])
}

// return all rows
function all(client) {
    return new Promise(function (resolve, reject) {
        client.query('SELECT * FROM rsvps', function (err, response) {
            resolve(response.rows)
        })
    })
}

// function to delete row (cant use delete as function name because it is a reseverd word in JS)
function remove(client, eventID, userID) {
    client.query('DELETE FROM rsvps WHERE linked_event_id = $1 AND linked_user_id = $2', [eventID, userID]);
}

module.exports = {
    'create': create,
    'retrieve': retrieve,
    'update': update,
    'all': all,
    'remove': remove
}

src/models/users.js
const apiError = require('../errors/apiError');

class User {

    constructor(userid, firstName, lastName, username, email) {
        this.id = userid;
        this.firstName = firstName;
        this.lastName = lastName;
        this.username = username;
        this.email = email;
        this.password
    }

    output() {
        return {
            "data": {
                "id": this.id,
                "firstName": this.firstName,
                "lastName": this.lastName,
                "username": this.username,
                "email": this.email
            }
        }
    }

    update(client, data) {
        client.query('UPDATE users SET first_name = $1, last_name = $2, username = $3, email = $4, user_password = $5 WHERE userid = $6', [data.firstName, data.lastName, data.username, data.email, data.password, data.userid])
    }

    all(client) {
        return new Promise(function (resolve, reject) {
            client.query('SELECT * FROM users', function (err, response) {
                resolve(response.rows)
            })
        })
    }

    remove(client) {
        client.query('DELETE FROM users WHERE userid = $1', [this.id]);
    }

}

function create(client, data) {

    return new Promise(function (resolve, reject) {
        client.query('INSERT INTO users (first_name, last_name, username, email, user_password) VALUES ($1, $2, $3, $4, $5) RETURNING userid', [data.firstName, data.lastName, data.username, data.email, data.password]).then((res) => {

            var user = new User(res.rows[0].userid, data.firstName, data.lastName, data.username, data.email, data.password)

            resolve(user);

        }).catch((err) => {
            console.log(err)
            reject(new apiError(parseInt(err.code), err.detail));
        })
    })
}

function retrieve(client, id) {
    return new Promise(function (resolve, reject) {
        client.query('SELECT * FROM users WHERE userid = $1', [id], function (err, response) {

            var data = response.rows[0]
            var user = new User(id, data.first_name, data.last_name, data.username, data.email, data.password)

            console.log(user);

            resolve(user)
        })
    })
}

function validate(data) {

    var cleanInputs = {};
    var keys = Object.keys(data);
    var values = Object.values(data);

    const re = /^(([^<>()[\]\\.,;:\s@\"]+(\.[^<>()[\]\\.,;:\s@\"]+)*)|(\".+\"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;

    // check email address
    if (!re.test(data.email)) {
        throw new Error('Invalid email address');
    }

    // check password
    if (data.password.length < 6) {
        throw new Error('Invalid password');
    }

    // trim extraneous chars and encode to UTF-8
    for (var i = 0; i < keys.length; i++) {
        cleanInputs[keys[i]] = values[i].trim();
    }

    return cleanInputs;

}

module.exports = {
    User,
    "validate": validate,
    "retrieve": retrieve,
    "create": create
}



src/models/events.js
// check if string is ISO-8601 format 
function isIsoDate(str) {
    if (!/\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}.\d{3}Z/.test(str)) return false;
    var d = new Date(str);
    return d.toISOString() === str;
}

// create an event
function create(client, data) {
    client.query('INSERT INTO events (event_name, event_description, event_timestamp) VALUES ($1, $2, $3)', [data.name, data.description, data.timestamp]);
}

// return event with specific id
function retrieve(client, id) {
    return new Promise(function (resolve, reject) {
        client.query('SELECT * FROM events WHERE event_id = $1', [id], function (err, response) {
            resolve(response.rows)
        })
    })
}

// update event with specific id
function update(client, data) {
    client.query('UPDATE events SET event_name = $1, event_description = $2, event_timestamp = $3 WHERE event_id = $5', [data.name, data.description, data.timestamp, data.event_id])
}

// return all rows
function all(client) {
    return new Promise(function (resolve, reject) {
        client.query('SELECT * FROM events', function (err, response) {
            resolve(response.rows)
        })
    })
}

function upcoming(client) {
    return new Promise(function (resolve, reject) {
        client.query('SELECT * FROM events WHERE event_timestamp >= CURRENT_TIMESTAMP', function (err, response) {
            resolve(response.rows)
        })
    })
}

// function to delete row (cant use delete as function name because it is a reseverd word in JS)
function remove(client, id) {
    client.query('DELETE FROM events WHERE event_id = $1', [id]);
}

// format (and decode) data for the frontend
function format(data) {
    var frontendData = {};

    // decode data and form frontend data obj
    frontendData.eventId = data.event_id;
    frontendData.name = data.event_name;
    frontendData.description = data.event_description;
    frontendData.timestamp = data.event_timestamp;

    return frontendData;
}

// validate (trim and encode) data
function validate(data) {

    var cleanInputs = {};
    var keys = Object.keys(data);
    var values = Object.values(data);

    // check timestamp
    if (!isIsoDate(data.timestamp)) {
        console.log('error')
        throw new Error('Invalid date and/or time');
    }
    
    // trim extraneous chars and encode to UTF-8
    for (var i = 0; i < keys.length; i++) {
        if (keys[i] != 'timestamp') {
            cleanInputs[keys[i]] = values[i].trim();
        } else {
            cleanInputs[keys[i]] = values[i].trim();
        }
    }

    return cleanInputs;

}

module.exports = {
    'create': create,
    'retrieve': retrieve,
    'update': update,
    'all': all,
    'upcoming': upcoming,
    'remove': remove,
    'format': format,
    'validate': validate
}

src/models/queries.js
// add event to DB
function addEvent(client, data) {
    client.query('INSERT INTO events (event_name, event_description, event_date, event_time) VALUES ($1, $2, $3, $4)', [data.name, data.description, data.date, data.time])
}

function returnEvent(client) {
    return new Promise(function (resolve, reject) {
        client.query('SELECT * FROM events', function (err, response) {
            if (err) {
                reject(err)
            } else {
                resolve(response.rows)
            }
        })
    })
}

// add RSVP to DB
function addRSVP(client, data) {
    client.query('INSERT INTO rsvps (linked_event_id, linked_event_name, linked_user_id, linked_username, rsvp_status) VALUES ($1, $2, $3, $4, $5)', [data.eventID, data.eventName, data.userID, data.username, data.rsvpStatus])
}

module.exports = {
    'addEvent': addEvent,
    'returnEvent': returnEvent,
    'addRSVP': addRSVP
}

src/errors/apiError.js
var errMessage;
var httpCode;

class apiError extends Error {
    constructor(errCode, msg) {
        switch (errCode) {
            case 300:
                errMessage = "Your request could not be completed because it was not properly authenticated.";
                httpCode = 401;
                break;

            case 301:
                errMessage = "A valid access token is required to access this resource.";
                httpCode = 401;
                break;

            case 302:
                errMessage = "You must be authenticated as a user to take this action.";
                httpCode = 401;
                break;

            case 310:
                errMessage = "The provided access token is invalid, or cannot be used with the provided parameters.";
                httpCode = 401;

            case 311:
                errMessage = "The provided refresh token is invalid, or cannot be used with the provided parameters.";
                httpCode = 400;
                break;

            case 312:
                errMessage = "The provided access token or refresh token does not match the given bearer token.";
                httpCode = 401;
                break;

            case 320:
                errMessage = "The provided refresh token cannot be used to renew an access token. Client login should be attempted.";
                httpCode = 403;
                break;

            case 700:
                errMessage = "The requested operation is not permitted on this resource.";
                httpCode = 412;
                break;

            case 701:
                errMessage = msg;
                httpCode = 412;
                break;

            case 23505:
                errMessage = msg;
                httpCode = 409;
                break;

            default:
                errCode = 303
                errMessage = "There was an error connecting to the server";
                httpCode = 502;
                break;
        }
        super(errMessage);
        this.errCode = errCode;
        this.httpCode = httpCode;
    }

    output() {
        return {
            "error": {
                "message": this.message,
                "errCode": this.errCode,
                "httpCode": this.httpCode
            }
        }

    }

}

module.exports = apiError;